### PR TITLE
fix: Find eigenvectors of defective matrices

### PR DESCRIPTION
**Pull request by gwhitney**
Previously, attempting to take the `eigs` of any defective matrix
  was doomed to fail in an attempt to solve a singular linear system.
  This PR detects the situation (as best as it can given the
  inherent numerical instability of the current methods used) and
  handles it. Note that in such cases, it's not possible to return
  a square matrix whose columns are the eigenvectors corresponding to
  the returned eigenvalues. In light of that fact and issue https://github.com/josdejong/mathjs/issues/3014, this
  PR also changes the return value of `eigs` so that the eigenvectors
  are passed back in a property `eigenvectors` which is an array of
  plain objects `{value: e, vector: v}`.

  Note that this PR makes the ancillary changes of correcting the
  spelling of the filename which was "realSymetric.js," and replacing
  the now-unnecessary auxiliary function "createArray" therein with
  `Array(size).fill(element)`. The rationale for performing these
  changes not strictly related to the issues at hand is that this
  file is rarely touched and with the level of maintenance hours we have
  at hand, it's more efficient to do these small refactorings in parallel
  with the actual bugfixes, which are orthogonal and so will not be
  obfuscated by this refactor. Note `git diff` does properly track the
  file name change.

  However, it also makes a potentially more pervasive change: in order for
  the numerically-sensitive algorithm to work, it changes the condition
  on when two very close (double) numbers are "nearlyEqual" from differing by
  less than DBL_EPSILON to differing by less than or equal to DBL_EPSILON.
  Although this may change other behaviors than the ones primarily being
  addressed, I believe it is an acceptable change because

  (a) It preserves all tests.
  (b) DBL_EPSILON is well below the standard config.epsilon anyway
  (c) I believe there are extant issues noting the odd/inconsistent
      behavior of nearlyEqual near 0 anyway, so I believe this will
      be overhauled in the future in any case. If so, the eigenvector
      computation will make a good test that a future nearlyEqual
      algorithm is working well.

  To be clear, the direct motivation for the change is that there are
  multiple cases in the eigenvector computation in which a coefficient
  that is "supposed" to be zero comes out to precisely DBL_EPSILON, which
  is fairly unsurprising given that these coefficients are produced by
  subtracting an eigenvalue from a diagonal entry of a matrix, which is
  likely to be essentially equal to that eigenvalue.

  As many tests of defective matrices as I could readily find by web
  searching have been added as unit tests (and one more in the typescript
  type testing). An additional case I found still fails, but in the
  _eigenvalue_ computation rather than the _eigenvector_ search, so that
  was deemed beyond the scope of this PR and has been filed as issue https://github.com/josdejong/mathjs/issues/3036.

  Resolves https://github.com/josdejong/mathjs/issues/2879.
  Resolves https://github.com/josdejong/mathjs/issues/2927.
  Resolves https://github.com/josdejong/mathjs/issues/3014.

### CODE REVIEW

**Comment by gwhitney** [comment]
Looks like I need to add tests in which the precision argument is used to pass the codecov/patch target. I will try that now.

**Comment by gwhitney** [comment]
Oh, and I forgot one more excuse/justification for changing nearlyEqual from `< DBL_EPSILON` to `<= DBL_EPSILON`:
(d) this is a breaking change anyway, so being slightly more relaxed in nearlyEqual, which could potentially be a breaking change even though we have no tests that detect that, should be more acceptable than in a regular non-breaking PR.

File: src/function/matrix/eigs.js [hunk 1]
```diff
@@ -41,63 +54,69 @@ export const createEigs = /* #__PURE__ */ factory(name, dependencies, ({ config,
...
       }
     }
-
-    const type = coerceTypes(mat, arr, N)
-    return doComplexEigs(arr, N, prec, type)
+    if (!result) {
+      const type = coerceTypes(mat, arr, N)
+      result = doComplexEigs(arr, N, prec, type)
+    }
+    if (matricize) {
```
Comment (josdejong) [inline src/function/matrix/eigs.js:107]: I'm personally not a big fan of having a mutable variable like `result` that is changed inside nested if's a few times, it makes it hard to follow the logic of what states `result` may have. 

Is it possible to extract this logic in two helper functions `matricize` and `attachVectorsError`, or a single one `prepareOutput` or so, and use that like `return prepareOutput(matricize(computeValuesAndVectors(mat, undefined)))`? Then you could keep this plumbing out of `computeValuesAndVectors` and basically keep it as it was with a direct `return` of the result.

File: src/function/matrix/eigs.js [hunk 2]
```diff
@@ -41,63 +54,69 @@ export const createEigs = /* #__PURE__ */ factory(name, dependencies, ({ config,
...
+    if (asize.length !== 2 || asize[0] !== asize[1]) {
+      throw new RangeError(`Matrix must be square (size: ${format(asize)})`)
     }
 
-    const arr = mat.toArray()
-    const N = size[0]
+    const N = asize[0]
 
     if (isReal(arr, N, prec)) {
       coerceReal(arr, N)
```
Comment (josdejong) [inline src/function/matrix/eigs.js:96]: I know this code was already there, so no regression, but I think `coerceReal(arr, N)` can mutate the original input data. I think we should clone the data first before mutating it.

File: src/function/matrix/eigs/complexEigs.js [hunk 3]
```diff
@@ -514,19 +498,17 @@ export function createComplexEigs ({ addScalar, subtract, flatten, multiply, mul
...
-
+    // compute diagonal elements of N = A - λI
     const na = subtract(a, l1)
-    const nb = subtract(b, l1)
-    const nc = subtract(c, l1)
     const nd = subtract(d, l1)
 
-    if (smaller(abs(nb), prec)) {
-      return [[na, one], [nc, zero]]
+    // N⃗₂ = 0 ⇒ S = ( N⃗₁, I⃗₁ )
```
Comment (josdejong) [inline src/function/matrix/eigs/complexEigs.js:505]: A thought: I think this file is the only one that contains non-ascii characters like `λ` and `N⃗₂ = 0 ⇒ S = ( N⃗₁, I⃗₁ )`. 

I haven't heard about issues with that in practice, but I can imagine that leads to weird cases. Maybe we should add a linter rule to check that we only use ascii characters to prevent possible issues in that regard. What do you think?

File: test/typescript-tests/testTypes.ts [hunk 4]
```diff
@@ -1282,6 +1282,17 @@ Matrices examples
     assert.strictEqual(math.matrix([1, 2, 3]) instanceof math.Matrix, true)
   }
 
+  // Eigenvalues and eigenvectors (should this be extended?)
```
Comment (josdejong) [inline test/typescript-tests/testTypes.ts:1285]: 👍 I think this covers the case, can you remove the question "should this be extended?"

File: test/unit-tests/function/matrix/eigs.test.js [hunk 5]
```diff
@@ -16,23 +16,36 @@ describe('eigs', function () {
...
+      assert(Array.isArray(genericArray.eigenvectors[ix].vector))
+    }
+    assert(genericArray.eigenvectors[0].vector[0] instanceof Complex)
 
     const realSymMatrix = eigs(matrix([[1, 0], [0, 1]]))
     assert(realSymMatrix.values instanceof Matrix)
     assert.deepStrictEqual(size(realSymMatrix.values), matrix([2]))
-    assert(realSymMatrix.vectors instanceof Matrix)
-    assert.deepStrictEqual(size(realSymMatrix.vectors), matrix([2, 2]))
+    for (let ix = 0; ix < 2; ++ix) {
```
Comment (josdejong) [inline test/unit-tests/function/matrix/eigs.test.js:34]: Would it be worth to create a little helper function to assert the output of `eigs`? The for loop is repeated quite a couple of times.

File: src/function/matrix/eigs/complexEigs.js [hunk 6]
```diff
@@ -444,30 +439,19 @@ export function createComplexEigs ({ addScalar, subtract, flatten, multiply, mul
       solutions.shift() // ignore the null vector
 
       // looks like we missed something, try inverse iteration
+      // But if that fails, just presume that the original matrix truly
+      // was defective.
       while (solutions.length < multiplicities[i]) {
```
Comment (josdejong) [inline src/function/matrix/eigs/complexEigs.js:444]: Question: I tried to see where the "actual" fix is for dealing with defective matrices but I'm not sure. Is it the removal of the logic around `failedLambdas` here?

**Review by josdejong — COMMENTED** [review]
Thanks a lot Glen, this looks good! I had expected that it would require more changes in the logic to fix this. 

I'm glad with your extra refactoring steps. I think the change in behavior around DBL_EPSILON is no problem.

Just good to mention: this PR contains a breaking change in the API of `eigs` and will be scheduled for `v12`.

File: src/function/matrix/eigs/complexEigs.js (hunk 6 above)
Comment (gwhitney) [reply src/function/matrix/eigs/complexEigs.js:444]: Three ingredients were necessary for the "actual" numerical fix: (1) accepting that if inverse iteration completely fails it's presumably because the matrix is actually defective (not a guarantee, but likely correct); (2) fixing the computation of the orthogonal component below; and (3) fixing the computation of the "jacobian2x2" which had a significant error in it.

File: src/function/matrix/eigs.js (hunk 2 above)
Comment (gwhitney) [reply src/function/matrix/eigs.js:96]: arr is the "working memory" for the algorithm, intended to be modified in place for efficiency. It was produced by mat.toArray(), so is already guaranteed to be unaliased and safe to modify in place. I will put a comment to that effect.

File: src/function/matrix/eigs.js (hunk 1 above)
Comment (gwhitney) [reply src/function/matrix/eigs.js]: OK, I _think_ I did the refactor you wanted. If this still isn't organized the way you like, please suggest another refactor and I am happy to make it so.

File: src/function/matrix/eigs/complexEigs.js (hunk 3 above)
Comment (gwhitney) [reply src/function/matrix/eigs/complexEigs.js]: OK, since these were bothering you, I converted the file to ascii (I think I got it all but didn't use any program to check).

But my two cents are: Unicode is explicitly allowed in JavaScript, and it is sometimes more compact/expressive, and the oldest PR that I want to get to back to as soon as "v13" is done will introduce the Unicode character √ into the read syntax for MEL (the mathjs expression language) , so at that point we'll be well and truly stuck with Unicode. So I don't see that it's worth anyone's effort to do anything to screen against it. If it passes the regression tests, I'm cool with unicode. I mean, some non-Roman non-Greek scripts might be difficult for _me_ to understand, but frankly I expect to live to a time when identifiers like 矢量 (according to the internet, means "vector" in Chinese) are commonplace in code packages, since I expect the world's fraction of Chinese-speaking/writing people will only increase...

File: test/typescript-tests/testTypes.ts (hunk 4 above)
Comment (gwhitney) [reply test/typescript-tests/testTypes.ts]: Done.

File: test/unit-tests/function/matrix/eigs.test.js (hunk 5 above)
Comment (gwhitney) [reply test/unit-tests/function/matrix/eigs.test.js]: Again, I think I've done the refactor you were hoping, but happy for advice on better versions...

**Comment by gwhitney** [comment]
OK, I think I've covered the review feedback, and I have incorporated an improvement to #2178 that I think is sufficient for the moment.

File: src/function/matrix/eigs.js (hunk 1 above)
Comment (josdejong) [reply src/function/matrix/eigs.js]: I like that 👍

File: src/function/matrix/eigs.js (hunk 2 above)
Comment (josdejong) [reply src/function/matrix/eigs.js]: I was assuming that `toArray` wouldn't copy the contents, but it actually does, so there is indeed no problem 😅 .

https://github.com/josdejong/mathjs/blob/develop/src/type/matrix/DenseMatrix.js#L665-L667

File: test/typescript-tests/testTypes.ts (hunk 4 above)
Comment (josdejong) [reply test/typescript-tests/testTypes.ts]: Thanks!

File: test/unit-tests/function/matrix/eigs.test.js (hunk 5 above)
Comment (josdejong) [reply test/unit-tests/function/matrix/eigs.test.js]: Thanks! Replacing the `for` loop with a `forEach` already makes it more compact and easy to follow, the helper function is a one-liner now.

File: src/function/matrix/eigs/complexEigs.js (hunk 6 above)
Comment (josdejong) [reply src/function/matrix/eigs/complexEigs.js:449]: O wow, thanks for the clarification.

**Comment by josdejong** [comment]
Thanks for the updates Glen, this looks good to go.

**Comment by josdejong** [comment]
I'll merge the PR into a new `v12` branch instead of `develop`.

**Comment by josdejong** [comment]
This fix has been published in `v12.0.0` now, thanks again Glen!
